runtime/src/prioritization_fee.rs

```diff
@@ -0,0 +1,354 @@
+                i64
+            ),
+            (
+                "total_prioritization_fee",
+                self.total_prioritization_fee as i64,
+                i64
+            ),
+            (
+                "total_update_elapsed_us",
+                self.total_update_elapsed_us as i64,
+                i64
+            ),
+        );
+    }
+}
+
+pub enum PrioritizationFeeError {
+    // Not able to get account locks from sanitized transaction, which is required to update block
+    // minimum fees.
+    FailGetTransactionAccountLocks,
+
+    // Not able to read priority details, including compute-unit price, from transaction.
+    // Compute-unit price is required to update block minimum fees.
+    FailGetTransactionPriorityDetails,
+
+    // Block is already finalized, trying to finalize it again is usually unexpected
+    BlockIsAlreadyFinalized,
+}
+
+/// Block minimum prioritization fee stats, includes the minimum prioritization fee of transactions in a
+/// block; and minimum fee for each writable accounts of all transactions in block. The only relevant
+/// write account minimum fees are those greater than block minimum fee, because the minimum fee needed to land
+/// a transaction is determined by Max( min_transaction_fee, min_writable_account_fees(key), ...)
+#[derive(Debug)]
+pub struct PrioritizationFee {
+    // The minimum prioritization fee of transactions that landed in this block.
+    min_transaction_fee: u64,
+
+    // The minimum prioritization fee of each writable account in transactions in this block.
+    min_writable_account_fees: HashMap<Pubkey, u64>,
+
+    // Default to `false`, set to `true` when a block is completed, therefore the minimum fees recorded
+    // are finalized, and can be made available for use (e.g., RPC query)
+    is_finalized: bool,
+
+    // slot prioritization fee metrics
+    metrics: PrioritizationFeeMetrics,
+}
+
+impl Default for PrioritizationFee {
+    fn default() -> Self {
+        PrioritizationFee {
+            min_transaction_fee: u64::MAX,
+            min_writable_account_fees: HashMap::new(),
+            is_finalized: false,
+            metrics: PrioritizationFeeMetrics::default(),
+        }
+    }
+}
+
+impl PrioritizationFee {
+    /// Use `sanitized_tx` to update self for minimum transaction fee in the block and minimum fee for each writable account.
+    pub fn update(
+        &mut self,
+        sanitized_tx: &SanitizedTransaction,
+    ) -> Result<(), PrioritizationFeeError> {
+        let (_, update_time) = measure!(
+            {
+                let account_locks = sanitized_tx
+                    .get_account_locks()
+                    .or(Err(PrioritizationFeeError::FailGetTransactionAccountLocks))?;
+
+                let priority_details = sanitized_tx
+                    .get_transaction_priority_details()
+                    .ok_or(PrioritizationFeeError::FailGetTransactionPriorityDetails)?;
+
```

Comment:
This may show my lack of Rust knowledge, but why does line 119 have "Err" in it, and this line not?

Reply:
Ln 119, `get_account_locks() ->Result<>`, `or` operator takes `Result` type; 
Ln 123, `get_transaction_priority_details() -> Option<>`, `ok_or` translate Option to Result